crates/ruma-client-api/src/threads/get_threads.rs

```diff
@@ -0,0 +1,113 @@
+//! `GET /_matrix/client/*/rooms/{roomId}/threads`
+//!
+//! Retrieve a list of threads in a room, with optional filters.
+
+pub mod v1 {
+    //! `/v1/` ([spec])
+    //!
+    //! [spec]: https://spec.matrix.org/v1.4/client-server-api/#get_matrixclientv1roomsroomidthreads
+
+    use js_int::UInt;
+    use ruma_common::{
+        api::ruma_api,
+        events::AnyMessageLikeEvent,
+        serde::{Raw, StringEnum},
+        RoomId,
+    };
+
+    use crate::PrivOwnedStr;
+
+    ruma_api! {
+        metadata: {
+            description: "Retrieve a list of threads in a room, with optional filters.",
+            method: GET,
+            name: "get_thread_roots",
+            unstable_path: "/_matrix/client/unstable/org.matrix.msc3856/rooms/:room_id/threads",
+            stable_path: "/_matrix/client/v1/rooms/:room_id/threads",
+            rate_limited: true,
+            authentication: AccessToken,
+            added: 1.4,
+        }
+
+        request: {
+            /// The room ID where the thread roots are located..
+            #[ruma_api(path)]
+            pub room_id: &'a RoomId,
+
+            /// The pagination token to start returning results from.
+            ///
+            /// If `None`, results start at the most recent topological event visible to the user.
+            #[serde(skip_serializing_if = "Option::is_none")]
+            #[ruma_api(query)]
+            pub from: Option<&'a str>,
+
+            /// Which thread roots are of interest to the caller.
+            #[serde(skip_serializing_if = "ruma_common::serde::is_default")]
+            #[ruma_api(query)]
+            pub include: IncludeThreads,
+
+            /// The maximum number of results to return in a single `chunk`.
+            ///
+            /// Servers should apply a default value, and impose a maximum value to avoid resource
+            /// exhaustion.
+            #[serde(skip_serializing_if = "Option::is_none")]
+            #[ruma_api(query)]
+            pub limit: Option<UInt>,
+        }
+
+        response: {
+            /// The thread roots, ordered by the `latest_event` in each event’s aggregation bundle.
+            ///
+            /// All events returned include bundled aggregations.
+            pub chunk: Vec<Raw<AnyMessageLikeEvent>>,
```

Comment:
The spec shows there being an optional `state_key`, so it seems like state events are allowed here? Might be a spec bug. If not this should be

```suggestion
            pub chunk: Vec<Raw<AnyTimelineEvent>>,
```

Reply:
I doubt it's a spec bug. It's the definition used in the MSC too, and there's also this in the spec about threads reply fallback:

> The most recent message known to the client in the thread.
> This should be something with a high chance of being rendered by the other client,
> such as an `m.room.message` event.

That seems to say there can also be events that are not usually rendered by other clients in the thread, which are usually state events.

Besides I've already seen in discussions that state events can have a relationship (even if it's not used in practice).